crates/oxc_minifier/src/compressor/fold.rs

```diff
@@ -180,4 +194,265 @@ impl<'a> Compressor<'a> {
+                        let number_literal = self.hir.number_literal(
+                            unary_expr.span,
+                            value,
+                            raw,
+                            NumberBase::Decimal,
+                        );
+                        Some(self.hir.literal_number_expression(number_literal))
+                    } else {
+                        None
+                    }
+                }
+            },
+            UnaryOperator::UnaryNegation => match &unary_expr.argument {
+                Expression::NumberLiteral(number_literal) => {
+                    let value = -number_literal.value;
+                    let raw = self.hir.new_str(value.to_string().as_str());
+                    let number_literal =
+                        self.hir.number_literal(unary_expr.span, value, raw, number_literal.base);
+                    Some(self.hir.literal_number_expression(number_literal))
+                }
+                Expression::Identifier(ident) => {
+                    if ident.name == "NaN" {
+                        self.try_detach_unary_op(unary_expr)
+                    } else {
+                        None
+                    }
+                }
+                _ => None,
+            },
+            UnaryOperator::BitwiseNot => {
+                if let Expression::NumberLiteral(number_literal) = &unary_expr.argument {
+                    let value = number_literal.value;
+                    if value.fract() == 0.0 {
+                        let int_value = ecmascript_to_int32(value);
+                        let number_literal = self.hir.number_literal(
+                            unary_expr.span,
+                            f64::from(!int_value),
+                            number_literal.raw,
+                            number_literal.base,
+                        );
+                        Some(self.hir.literal_number_expression(number_literal))
+                    } else {
+                        None
+                    }
+                } else {
+                    None
+                }
+            }
+            _ => None,
+        }
+    }
+
+    fn try_detach_unary_op(
+        &mut self,
+        unary_expr: &mut UnaryExpression<'a>,
+    ) -> Option<Expression<'a>> {
+        match &unary_expr.argument {
+            Expression::Identifier(ident) => {
+                if matches!(ident.name.as_str(), "NaN" | "Infinity") {
+                    let ident = self.hir.identifier_reference(
+                        unary_expr.span,
+                        ident.name.clone(),
+                        ident.reference_id,
+                    );
+                    Some(self.hir.identifier_reference_expression(ident))
+                } else {
+                    None
+                }
+            }
+            _ => None,
+        }
+    }
+}
+
+fn get_number_value(expr: &Expression) -> Option<f64> {
```

Comment:
It confuses me where I should put it when it not a fold function? I put it at the blow of this file. Is that a good way?

Reply:
Make it a ast method, we also have the `oxc_syntax` crate